homeassistant/components/toon/config_flow.py

```diff
@@ -1,166 +1,92 @@
+        )
         try:
-            toon = await self.hass.async_add_executor_job(
-                partial(
-                    Toon,
-                    user_input[CONF_USERNAME],
-                    user_input[CONF_PASSWORD],
-                    app[CONF_CLIENT_ID],
-                    app[CONF_CLIENT_SECRET],
-                    tenant_id=user_input[CONF_TENANT],
-                )
-            )
-
-            displays = toon.display_names
-
-        except InvalidConsumerKey:
-            return self.async_abort(reason=CONF_CLIENT_ID)
-
-        except InvalidConsumerSecret:
-            return self.async_abort(reason=CONF_CLIENT_SECRET)
+            self.agreements = await toon.agreements()
+        except ToonError:
+            return self.async_abort(reason="connection_error")
 
-        except InvalidCredentials:
-            return await self._show_authenticaticate_form({"base": "credentials"})
-
-        except AgreementsRetrievalError:
+        if not self.agreements:
             return self.async_abort(reason="no_agreements")
 
-        except Exception:  # pylint: disable=broad-except
-            _LOGGER.exception("Unexpected error while authenticating")
-            return self.async_abort(reason="unknown_auth_fail")
-
-        self.displays = displays
-        self.username = user_input[CONF_USERNAME]
-        self.password = user_input[CONF_PASSWORD]
-        self.tenant = user_input[CONF_TENANT]
+        return await self.async_step_agreement()
 
-        return await self.async_step_display()
+    async def async_step_import(self, config: Dict[str, Any]) -> Dict[str, Any]:
+        """Start a configuration flow based on imported data.
 
-    async def _show_display_form(self, errors=None):
-        """Show the select display form to the user."""
-        fields = OrderedDict()
-        fields[vol.Required(CONF_DISPLAY)] = vol.In(self.displays)
-
-        return self.async_show_form(
-            step_id="display",
-            data_schema=vol.Schema(fields),
-            errors=errors if errors else {},
-        )
+        This step is merely here to trigger "discovery" when the `toon`
+        integration is listed in the user configuration, or when migrating from
+        the version 1 schema.
+        """
+        await self._async_handle_discovery_without_unique_id()
+        return await self.async_step_user()
 
-    async def async_step_display(self, user_input=None):
-        """Select Toon display to add."""
+    async def async_step_agreement(
+        self, user_input: Dict[str, Any] = None
+    ) -> Dict[str, Any]:
+        """Select Toon agreement to add."""
+        if len(self.agreements) == 1:
+            return await self._create_entry(self.agreements[0])
 
-        if not self.displays:
-            return self.async_abort(reason="no_displays")
+        agreements_list = [
+            f"{agreement.agreement_id} - {agreement.street} {agreement.house_number}, {agreement.city}"
```

Comment:
Should we filter out the agreements that are already configured for this account ?

Reply:
Also, is there a reason we would only allow 1 agreement to be connected to 1 config entry ? If it is all the same auth, shouldn't we just activate all agreements with maybe a picker in options dialog to enable/disable agreements ?

Reply:
Having 2 or more agreements would be really really rare. It basically would mean you have multiple "houses" on your single energy contract. Even in that case, I think one should be able to select the address to add specifically (considering Home Assistant's nature to run in one home at the time).

Nevertheless, it leaves the possibility open, but I feel like I should not put too much effort into it.

Reply:
Ok!